strings/bitap_string_match.py

```diff
@@ -0,0 +1,79 @@
+"""
+Bitap exact string matching
+https://en.wikipedia.org/wiki/Bitap_algorithm
+
+Searches for a pattern inside text, and returns the index of the first occurence
+of the pattern. Both text and pattern consist of lowercase alphabetical characters only.
+
+Complexity: O(m*n)
+    n = length of text
+    m = length of pattern
+
+Python doctests can be run using this command:
+python3 -m doctest -v bitap_string_match.py
+"""
+
+
+def bitap_string_match(text: str, pattern: str) -> int | None:
+    """
+    Retrieves the index of the first occurrence of pattern in text.
+
+    Args:
+        text: A string consisting only of lowercase alphabetical characters.
+        pattern: A string consisting only of lowercase alphabetical characters.
+
+    Returns:
+        int: The index where pattern first occurs.
+
+    >>> bitap_string_match('abdabababc', 'ababc')
+    5
+    >>> bitap_string_match('aaaaaaaaaaaaaaaaaa', 'a')
+    0
+    >>> bitap_string_match('zxywsijdfosdfnso', 'zxywsijdfosdfnso')
+    0
+    >>> bitap_string_match('abdabababc', '')
+    0
+    >>> bitap_string_match('abdabababc', 'c')
+    9
+    >>> bitap_string_match('abdabababc', 'fofosdfo') is None
+    True
+    >>> bitap_string_match('abdab', 'fofosdfo') is None
+    True
+    """
+    m: int = len(pattern)
+    if m == 0:
+        return 0
+    if m > len(text):
+        return None
```

Comment:
I am not a fan of functions that can return multiple data types because they are merely pushing work onto the caller.  Either raise a ValueError() like `str.index()` or return a `-1` like `str.find()`
* https://docs.python.org/3/library/stdtypes.html#str.index
* https://docs.python.org/3/library/stdtypes.html#str.find